File: leetcode/14D Algo2/sliding window/438. Find All Anagrams in a String.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<list>
#include<stack>
#include<unordered_set> 
#include<unordered_map> 
using namespace std;

/*
Given two strings s and p, return an array of all the 
start indices of p's anagrams in s. You may return the answer in any order.

An Anagram is a word or phrase formed by rearranging 
the letters of a different word or phrase, typically using all the original letters exactly once.

Input: s = "cbaebabacd", p = "abc"
Output: [0,6]
Explanation:
The substring with start index = 0 is "cba", which is an anagram of "abc".
The substring with start index = 6 is "bac", which is an anagram of "abc".
*/


class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> res(0);
        if(p.size() > s.size()) return res;
        int len = p.size();
        vector<char> pd(26, 0);
        vector<char> sd(26, 0);
        int i = 0;
        for(i; i < p.size(); ++i){
            ++pd[p[i] - 'a'];
            ++sd[s[i] - 'a'];
        }
        if(pd == sd){
            res.push_back(0);
        }
        while(i < s.size()){
            --sd[s[i - len] - 'a'];
            ++sd[s[i] - 'a'];
            if(pd == sd){
                res.push_back(i - len + 1);
            }  
            ++i;
        }
        return res;      
    }
}; 

File: leetcode/14D Algo2/sliding window/438. Find All Anagrams in a String.java
import java.util.*;

/*
Given two strings s and p, return an array of all 
the start indices of p's anagrams in s. You may return 
the answer in any order.

Input: s = "cbaebabacd", p = "abc"
Output: [0,6]
*/

class Solution {

    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new LinkedList<>();
        if(s.length() < p.length()) return res; 
        int[] Ms = new int[26]; int[] Mp = new int[26]; 
        for (int i = 0; i < p.length(); i++) {
            ++Ms[s.charAt(i) - 'a'];
            ++Mp[p.charAt(i) - 'a'];
        }
        if(Arrays.equals(Ms, Mp)) res.add(0);
//         --------------
        int l = 1, r = p.length();
        while(r < s.length()){
            --Ms[s.charAt(l - 1) - 'a'];
            ++Ms[s.charAt(r) - 'a'];
            if(Arrays.equals(Ms, Mp)){
//             --------------
                res.add(l);
            }
            ++l;
            ++r;
        }
        return res;
    }
}